drivers/bus/isapnp/pdo.c

```diff
@@ -183,146 +524,381 @@ IsaPdoQueryResourceRequirements(
+        DPRINT1("Bad resource list version (%d.%d)\n",
+                ResourceList->List[0].PartialResourceList.Version,
+                ResourceList->List[0].PartialResourceList.Revision);
         return STATUS_REVISION_MISMATCH;
     }
-    for (i = 0; i < ResourceList->List[0].PartialResourceList.Count; i++)
+
+#if 0
+    /* Try various Read Ports from the list */
+    if (ResourceList->List[0].PartialResourceList.Count > 3)
     {
-        PCM_PARTIAL_RESOURCE_DESCRIPTOR PartialDescriptor = &ResourceList->List[0].PartialResourceList.PartialDescriptors[i];
-        if (PartialDescriptor->Type == CmResourceTypePort && PartialDescriptor->u.Port.Length > 1 && !FdoExt->ReadDataPort)
+        for (i = 0; i < ResourceList->List[0].PartialResourceList.Count; i++)
         {
-            PUCHAR ReadDataPort = ULongToPtr(PartialDescriptor->u.Port.Start.u.LowPart + 3);
-            if (NT_SUCCESS(IsaHwTryReadDataPort(ReadDataPort)))
+            PCM_PARTIAL_RESOURCE_DESCRIPTOR PartialDescriptor =
+                &ResourceList->List[0].PartialResourceList.PartialDescriptors[i];
+
+            if (IS_READ_PORT(PartialDescriptor))
             {
-                /* we detected some ISAPNP cards */
-                FdoExt->ReadDataPort = ReadDataPort;
-                KeAcquireSpinLock(&FdoExt->Lock, &OldIrql);
-                Status = IsaHwFillDeviceList(FdoExt);
-                KeReleaseSpinLock(&FdoExt->Lock, OldIrql);
-                if (FdoExt->DeviceCount > 0)
+                PUCHAR ReadDataPort = ULongToPtr(PartialDescriptor->u.Port.Start.u.LowPart + 3);
+                ULONG Cards;
+
+                Cards = IsaHwTryReadDataPort(ReadDataPort);
+                IsaHwWaitForKey();
+
+                /* We detected some ISAPNP cards */
+                if (Cards > 0)
                 {
-                    IoInvalidateDeviceRelations(FdoExt->Pdo, BusRelations);
-                    IoInvalidateDeviceRelations(FdoExt->DataPortPdo, RemovalRelations);
+                    if (PdoExt->RequirementsList)
+                        ExFreePoolWithTag(PdoExt->RequirementsList, TAG_ISAPNP);
+
+                    /* Discard the Read Ports at conflicting locations */
+                    Status = IsaPnpCreateReadPortDORequirements(PdoExt,
+                                                                PartialDescriptor->
+                                                                u.Port.Start.u.LowPart);
+                    if (!NT_SUCCESS(Status))
+                        return Status;
+
+                    PdoExt->Flags |= ISAPNP_READ_PORT_NEED_REBALANCE;
+
+                    IoInvalidateDeviceState(PdoExt->Common.Self);
+
+                    return STATUS_SUCCESS;
                 }
             }
-            else
+        }
+    }
+    /* Set the Read Port */
+    else if (ResourceList->List[0].PartialResourceList.Count == 3)
+#else
+    if (ResourceList->List[0].PartialResourceList.Count > 3) /* Temporary HACK */
```

Comment:
Is this the hack you've mentioned in the PR comments?

Reply:
Yes